tracks/bash/exercises/hamming/mentoring.md

```diff
@@ -0,0 +1,135 @@
+### Reasonable Solutions
+```bash
+#!/usr/bin/env bash
+main () {
+  (( "$#" == 2 )) || {
+      echo "Usage: hamming.sh <string1> <string2>"
+      exit 1
+  }
+  
+  # Regular vars are easier to read when doing fancy parameter expansion.
+  a=$1 b=$2 
+  
+  (( "${#a}" == "${#b}" )) || {
+      echo "left and right strands must be of equal length"
+      exit 1
+  }
+
+  declare -i count
+  for (( i = 0; i < "${#a}"; i++ )); do
+    [[ "${a:i:1}" == "${b:i:1}" ]] || count+=1
+  done
+
+  printf '%d\n' "$count"
+}
+
+main "$@"
+```
+```bash
+#!/usr/bin/env bash
+main () {
+    if (( $# != 2 )); then 
+        echo "Usage: hamming.sh <string1> <string2>"
+        exit 1
+    fi
+    if (( ${#1} != ${#2} )); then
+        echo "left and right strands must be of equal length"
+        exit 1
+    fi
+    hamming=0
+    for (( i=0; i<${#1}; i++ )); do
+        [[ "${1:$i:1}" != "${2:$i:1}" ]] && (( ++hamming ))
+    done
+    echo "$hamming"
+}
+
+main "$@"
+```
+
+### Common Suggestions
+* Suggest using a for loop in bash or other [looping constructs](https://www.gnu.org/software/bash/manual/html_node/Looping-Constructs.html#Looping-Constructs) https://www.gnu.org/software/bash/manual/html_node/Looping-Constructs.html#Looping-Constructs
+
+* Suggest using [Parameter Expansion - Substring Extraction](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html) to access the index of the string. https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html
+
+* Suggest creating a counter to gather hammingDistance
+1) let "hammingDistance+=1"
+2) counter=0; (( ++counter ));
+3) counter=0; counter=(( $counter+1 ));
+
+* Suggest to quote the right-hand side of == in [[ ]] to prevent glob matching, the left hand side can be quoted too.
+```bash
+@test "expose subtle '[[ \$x == \$y ]]' bug" {
+  [[ $BATS_RUN_SKIPPED == "true" ]] || skip
+  run bash hamming.sh 'AAA' 'A?A'
+  (( status == 0 ))
+  [[ $output == "1" ]]
+}
+```
+output:
+```
+✗ expose subtle '[[ $x == $y ]]' bug
+   (in test file hamming_test.sh, line 92)
+     `(( status == 0 ))' failed
+```
+Problem line of code: 
+```bash
+a=$1
+b=$2
+[[ ${a:i:1} == ${b:i:1} ]] || count+=1
+```
+Potential fix:
+```bash
+a=$1
+b=$2
+[[ "${a:i:1}" == "${b:i:1}" ]] || count+=1
+```
+
+### Talking Points
+* List Constructs used instead of `if`/`else`
+
+The `&&` and `||` in bash are short-circuit logical operators. They quit evaluating as soon as possible.
```

Comment:
```suggestion
The `&&` and `||` in bash are short-circuit logical operators. They stop evaluating as soon as possible.
```

This matches the wording below where this idea is repeated in an example.

Reply:
added this change